scylla-rust-wrapper/src/collection.rs

```
@@ -57,18 +75,61 @@ pub unsafe extern "C" fn cass_collection_new(
+        // the number of CassCqlValue will be stored.
         CassCollectionType::CASS_COLLECTION_TYPE_MAP => item_count * 2,
         _ => item_count,
     } as usize;
 
     Box::into_raw(Box::new(CassCollection {
         collection_type,
+        data_type: None,
+        capacity,
+        items: Vec::with_capacity(capacity),
+    }))
+}
+
+#[no_mangle]
+unsafe extern "C" fn cass_collection_new_from_data_type(
+    data_type: *const CassDataType,
+    item_count: size_t,
+) -> *mut CassCollection {
+    let data_type = clone_arced(data_type);
+    let (capacity, collection_type) = match data_type.as_ref() {
+        CassDataType::List { .. } => (item_count, CassCollectionType::CASS_COLLECTION_TYPE_LIST),
+        CassDataType::Set { .. } => (item_count, CassCollectionType::CASS_COLLECTION_TYPE_SET),
+        // Maps consist of a key and a value, so twice
+        // the number of CassCqlValue will be stored.
+        CassDataType::Map { .. } => (item_count * 2, CassCollectionType::CASS_COLLECTION_TYPE_MAP),
+        _ => return std::ptr::null_mut(),
+    };
+    let capacity = capacity as usize;
+
+    Box::into_raw(Box::new(CassCollection {
+        collection_type,
+        data_type: Some(data_type),
         capacity,
         items: Vec::with_capacity(capacity),
     }))
 }
 
+#[no_mangle]
+unsafe extern "C" fn cass_collection_data_type(
+    collection: *const CassCollection,
+) -> *const CassDataType {
+    let collection_ref = ptr_to_ref(collection);
+
+    match &collection_ref.data_type {
+        Some(dt) => Arc::as_ptr(dt),
+        None => match collection_ref.collection_type {
+            CassCollectionType::CASS_COLLECTION_TYPE_LIST => &UNTYPED_LIST_TYPE,
+            CassCollectionType::CASS_COLLECTION_TYPE_SET => &UNTYPED_SET_TYPE,
+            CassCollectionType::CASS_COLLECTION_TYPE_MAP => &UNTYPED_MAP_TYPE,
+            // CassCollectionType is a C enum. Can user provide a value out of range?
+            _ => unreachable!(),
```

User can provide a value out of range, but then it's going to be their fault. That's how C contract on UB works...

---

I think we should panic here then (with some insightful error message). I wouldn't silently return a null pointer.